C++ APIs Transformer NN Module Top Layer

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44333 C++ APIs Transformer NN Module Top Layer**

This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

Commit:
C++ APIs Transformer NN Module Top Layer

[ghstack-poisoned]

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]

zou3519 commented on torch/csrc/api/include/torch/nn/options/transformer.h:
````diff
@@ -0,0 +1,57 @@
+
+/// Options for the `Transformer` module
+///
+/// Example:
+/// ```
+///
+/// ```
````
No example?

glaringlee replied:
Ah.....forgot to add it back....was doing cut, but forgot to paste back......will add

zou3519 commented on torch/csrc/api/include/torch/nn/options/transformer.h:
```diff
@@ -0,0 +1,57 @@
+  using activation_t = c10::variant<enumtype::kReLU, enumtype::kGELU>;
+
+  TransformerOptions() = default;
+  TransformerOptions(int64_t d_model, int64_t nhead);
+  TransformerOptions(int64_t d_model, int64_t nhead, int64_t num_encoder_layers, int64_t num_decoder_layers);
```
(question) Why did we choose to provide these two constructors and not, say,
```
TransformerOptions(int64_t d_model, int64_t nhead, int64_t num_encoder_layers, int64_t num_decoder_layers, dim_feedforward, dropout, activation)
```
?

glaringlee replied:
From the usage of python version, I found out that when people initialize the transformer, these two are mostly used. 
So I provide these two.

zou3519 replied:
Maybe add a comment to that effect in the code, so someone in the future doesn't go and try adding all of the overloads

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformercoder.h:
```diff
@@ -85,47 +85,48 @@ TORCH_MODULE(TransformerEncoder);
-
- protected:
-  FORWARD_HAS_DEFAULT_ARGS(
+
+  public:
+    TransformerDecoderImpl(TransformerDecoderLayer decoder_layer, int64_t num_layers)
+      : TransformerDecoderImpl(TransformerDecoderOptions(decoder_layer, num_layers)) {}
+    explicit TransformerDecoderImpl(TransformerDecoderOptions options_);
+
+    void reset() override;
+
+    void reset_parameters();
```
Are the spacing changes intentional? Do we have a PyTorch C++ API style guide?

glaringlee replied:
reset_parameters() has no regulation. Since this is not required for every module and basically a customized function. I am trying to make it always named reset_parameters(). I will enter another PR to change all _reset_parameters(), reset_parameters_() into reset_parameter(). 
We used to have a ticket to make this function virtual. But we did not proceed since this is not needed by every module and making function virtual will lower the performance.

zou3519 replied:
Sorry, my question is that it looks like there has been formatting changes. Previously, there were fewer spaces, now, there are more spaces. Is the addition of spaces intentional?

glaringlee replied:
I see. Yes, the decoder part is contributed from external. 
We follow the the convention that with 2 chars ident. For each level, so I corrected it for decoder impl.

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -80,78 +85,58 @@ TransformerDecoderLayerImpl::TransformerDecoderLayerImpl(
-  norm3 = register_module(
-    "norm3",
-    LayerNorm(LayerNormOptions(std::vector<int64_t> {options.d_model()})));
+  // NOTE: reset() is for initializing the model only, call reset() after the model is created
+  // will cause throwing exceptions. Call reset_parameter() if the created model need a reset.
```
nit: "call reset() after the model is created" -> "calling reset() after the model is created"
nit: "call reset_paramter if the created model need a reset" ->  "call reset_paramter if the created model needs a reset"

zhangguanheng66 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+  return output;
+}
+
+Tensor TransformerImpl::generate_square_subsequent_mask(int64_t sz) {
```
I'm not sure if we want to have `generate_square_subsequent_mask` func within the transformer. Although in the python version it is, it's really not part of the paper "Attention is All Your Need".

zou3519 replied:
@zhangguanheng66 do you know why `generate_square_subsequent_mask` is exposed as a part of the Python API?

zhangguanheng66 replied:
@zhangguanheng66 if I could start from scratch, I would not include it in the transformer.

glaringlee replied:
@zou3519 @zhangguanheng66 
If we know someone already use this function in python, then I think we'd better keep this function here since otherwise, this will be an unexpected bc breaking change to user.

zou3519 replied:
It's not really BC-breaking if the functionality doesn't exist in C++ yet. It sounds like we might want to deprecate `generate_square_subsequent_mask` on the Python side?

glaringlee replied:
@zhangguanheng66 
Should I deprecate this? Or I probably can move this to the util namespace.
But I need to know the following:
1.  Is this function used specifically for transformer? (If so, probably leave this here, otherwise, can put this in nn::util)
2. Is this a generic method to generate square subsequent masks?  
3. Any other place such mask will be used within or codebase?
Thanks

zhangguanheng66 replied:
This func is mainly used to generate a square masks for the transformer application. In the word language task, word in a sentence is only allowed to see the tokens before itself so we need an attention mask here. Can we leave this func in the transformer namespace but as an independent function, rather than part of the transformer model?

glaringlee replied:
@zou3519 @zhangguanheng66 
I am going to make this static method within Transformer class. People just call Transformer::generate_sequare_subsequent_mask() instead of creating an obj and invoke it from obj.

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+    this->encoder = AnyModule(trans_encoder);
+  }
+  else {
+    this->encoder = options.custom_encoder().clone();
```
.clone() does a deep copy, right?

zhangguanheng66 replied:
A deep copy is what we want here. Better to make sure that it made a deep copy here.

zou3519 replied:
https://github.com/pytorch/pytorch/blob/367426494759ddde0896665ed55c6f9af2870cf0/torch/csrc/api/include/torch/nn/module.h#L98-L115

Confirmed that it is a deep copy

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+
+
+void TransformerImpl::reset_parameters() {
+  auto parameters = this->parameters();
```
Should probably be `auto&`, you don't want to (shallow) copy the vector, https://github.com/pytorch/pytorch/blob/367426494759ddde0896665ed55c6f9af2870cf0/torch/csrc/api/include/torch/nn/module.h#L208

zou3519 replied:
Actually, nevermind, I see that there is no API to get a reference to the parameters like that

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+  TORCH_CHECK(src.dim() == 3 && tgt.dim() == 3,
+    "src and tgt should have 3 dimensions, but got ", src.dim(), " and ", tgt.dim());
+
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have same size at dimension 1, but got ", src.size(1), " and ", tgt.size(1));
```
nit: The python version mentions the "batch" dimension in the error message which is more readable IMO: https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py#L118

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have same size at dimension 1, but got ", src.size(1), " and ", tgt.size(1));
+
+  TORCH_CHECK(src.size(2) == options.d_model() && tgt.size(2) == options.d_model(),
+    "src and tgt should have a same size as d_model at dimension 2, but got ",
+    src.size(2), " and ", tgt.size(2), " while d_model is ", options.d_model());
```
Nit: the python version mentions "feature" dimension in the error message which is more readable IMO: https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py#L121, it would be nice to include that as well as talk about dim 2

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+  else {
+    TORCH_WARN_ONCE(
+      "IEEE754 is not supporetd on this platform, generate_square_subsequent_mask will fill "
+      "the mas with smallest float number on this platform instead of -inf");
```
"mas" -> "mask"

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+
+  Tensor mask = (torch::triu(torch::ones({sz, sz})) == 1).transpose(0, 1).to(torch::kFloat32);
+
+  // check IEEE754 support here since -inf is not guaranteed to be valid on non IEEE754 platform
+  if (std::numeric_limits<float>::is_iec559) {
+    mask = mask.masked_fill(mask == 0, -std::numeric_limits<float>::infinity()).masked_fill(mask == 1, 0.f);
+  }
+  // if IEEE754 is not supported, we use the smallest float number in current platform
```
This seems fine, but does the Python API have the same behavior?

glaringlee replied:
Python doesn't have this issue. 
I checked with @zhangguanheng66, using the smallest float number in this situation is fine.

zou3519 commented on test/cpp/api/transformer.cpp:
```diff
@@ -8,6 +8,16 @@ using namespace torch::nn;
 struct TransformerTest : torch::test::SeedingFixture {};
 
+// a generic function to set constants for parameters so we have fixed result for deterministic test
+template<typename M>
```
nit: prefer more descriptive typename (e.g., model_t, module_t)

zou3519 commented on test/cpp/api/transformer.cpp:
```diff
@@ -8,6 +8,16 @@ using namespace torch::nn;
 
+// a generic function to set constants for parameters so we have fixed result for deterministic test
+template<typename M>
+void set_parameter_constants(M& model, const torch::TensorOptions& tensor_options) {
```
nit: "set_parameter_to_constants" makes it a tad bit more readable

zou3519 reviewed (commented):
Overall this looks good to me, the two main questions I have are:
1) should `generate_square_subsequent_mask` be a part of the Transformer API? I don't know what it's for
2) Is there a better way to test this? It would be nice if we could use the parity harness; it's hard for a code reader to figure out correctness from reading the sample inputs/outputs

glaringlee commented:
> Overall this looks good to me, the two main questions I have are:
> 
> 1. should `generate_square_subsequent_mask` be a part of the Transformer API? I don't know what it's for
> 2. Is there a better way to test this? It would be nice if we could use the parity harness; it's hard for a code reader to figure out correctness from reading the sample inputs/outputs

1. See my comments to @zhangguanheng66 

2. Yes, parity will be turned on in next PR. This test is hard to read, since it is a deterministic test (I run the same thing in python and get the result, then run the c++ code and compare result with the python result).

I will address other comments later.

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]

glaringlee commented:
This update fix all previous comments except the `generate_square_subsequent_mask`. Will do the update once @zhangguanheng66 comment.

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]

glaringlee commented:
changed the `generate_square_subsequent_mask` as static. Expose a type alias `Impl` from Transformer class (due to the special implementation of Module, we can not use `Transformer::generate_square_subsequent_mask` directly. So I standardized this to `Transformer::Impl::generate_square_subsequent_mask`.

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -13,8 +18,8 @@ TransformerEncoderLayerImpl::TransformerEncoderLayerImpl(
-  // NOTE: reset() is for initializing the model only, call reset() after the model is created
-  // will cause throwing exceptions. Call reset_parameter() if the created model need a reset
+  // NOTE: reset() is for initializing the model only, calling reset() after the model is created
+  // will cause throwing exceptions. Call reset_parameter() if the created model needs a reset
```
nit: "will cause throwing exceptions" -> "will cause exceptions to be thrown", or "will throw exceptions"

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -80,78 +85,58 @@ TransformerDecoderLayerImpl::TransformerDecoderLayerImpl(
-    "norm3",
-    LayerNorm(LayerNormOptions(std::vector<int64_t> {options.d_model()})));
+  // NOTE: reset() is for initializing the model only, calling reset() after the model is created
+  // will cause throwing exceptions. Call reset_parameter() if the created model needs a reset.
```
nit: "will cause throwing exceptions" -> "will cause exceptions to be thrown", or "will throw exceptions"

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformer.h:
```diff
@@ -0,0 +1,140 @@
+    ///   2. This function requires the platform support IEEE754, since `-inf` is guaranteed to
+    ///      be valid only when IEEE754 is supported. If the platform doesn't support IEEE754,
+    ///      this function will fill the mask with the smallest float number instead of `-inf`,
+    ///      a one time warning will be pop up as well.
```
nit: "will be pop up as well" -> "will pop up as well"

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformer.h:
```diff
@@ -0,0 +1,140 @@
+/// model with corresponding parameters.
+///
+/// See https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html to
+/// learn abouut the exact behavior of this transformer model
```
nit: "abouut" -> "about"

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+    "src and tgt should have 3 dimensions, but got ", src.dim(), " and ", tgt.dim());
+
+  TORCH_CHECK(src.size(1) == tgt.size(1),
+    "src and tgt should have equal batch number (at dim 1), but got ", src.size(1), " and ", tgt.size(1));
```
nit: "batch size" or "batch dim size" sounds more pytorch-like than "batch number"

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -342,5 +326,110 @@ Tensor TransformerDecoderImpl::forward(
+    "src and tgt should have equal batch number (at dim 1), but got ", src.size(1), " and ", tgt.size(1));
+
+  TORCH_CHECK(src.size(2) == options.d_model() && tgt.size(2) == options.d_model(),
+    "src and tgt should have same feature number as d_model (at dim 2), but got ",
```
nit: "feature size" sounds more pytorch-like than "feature number"

zou3519 commented on test/cpp/api/transformer.cpp:
```diff
@@ -1036,54 +1040,129 @@ TEST_F(TransformerTest, PrettyPrintTransformerDecoder) {
+    ASSERT_EQ(result.sizes(), ref_output.sizes());
+    ASSERT_TRUE(result.equal(result_cus));
+    ASSERT_TRUE(torch::allclose(result, ref_output, 1e-7, 1e-5, /*equal_nan=*/true));
+}
```
I'd also test the error conditions (e.g., src and target don't have the same size in the batch dim) with an ASSERT_RAISES to check that it doesn't crash in an unfortunate way

glaringlee replied:
will add these tests

zou3519 reviewed (approved):
LGTM, some very minor comments

glaringlee commented:
> LGTM, some very minor comments

will address

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]

Commit:
Update on "C++ APIs Transformer NN Module Top Layer"


This is to provide C++ APIs for Transformer NN Module which has been implemented in python here:https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/transformer.py

It also adjusted the TransformerDecoderLayer's module registration order to 100% match the python impl.

Differential Revision: [D23584010](https://our.internmc.facebook.com/intern/diff/D23584010)

[ghstack-poisoned]
